driver-core/src/test/functional/com/mongodb/internal/connection/TlsChannelStreamFunctionalTest.java

```diff
@@ -0,0 +1,149 @@
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.internal.connection.OperationContext.simpleOperationContext;
+import static java.lang.String.format;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.verify;
+
+class TlsChannelStreamFunctionalTest {
+    private static final SslSettings SSL_SETTINGS = SslSettings.builder().enabled(true).build();
+    private static final String UNREACHABLE_PRIVATE_IP_ADDRESS = "10.255.255.1";
+    private static final int UNREACHABLE_PORT = 65333;
+
+    @ParameterizedTest
+    @ValueSource(ints = {500, 1000, 2000})
+    void shouldInterruptConnectionEstablishmentWhenConnectionTimeoutExpires(final int connectTimeoutMs) throws IOException {
+        //given
+        try (StreamFactoryFactory streamFactoryFactory = new TlsChannelStreamFactoryFactory(new DefaultInetAddressResolver());
+             MockedStatic<SocketChannel> socketChannelMockedStatic = Mockito.mockStatic(SocketChannel.class)) {
+            SingleResultSpyCaptor<SocketChannel> singleResultSpyCaptor = new SingleResultSpyCaptor<>();
+            socketChannelMockedStatic.when(SocketChannel::open).thenAnswer(singleResultSpyCaptor);
+
+            StreamFactory streamFactory = streamFactoryFactory.create(SocketSettings.builder()
+                    .connectTimeout(connectTimeoutMs, TimeUnit.MILLISECONDS)
+                    .build(), SSL_SETTINGS);
+
+            Stream stream = streamFactory.create(new ServerAddress(UNREACHABLE_PRIVATE_IP_ADDRESS, UNREACHABLE_PORT));
+            long connectOpenStart = System.nanoTime();
+
+            //when
+            OperationContext operationContext = createOperationContext(connectTimeoutMs);
+            MongoSocketOpenException mongoSocketOpenException = assertThrows(MongoSocketOpenException.class, () ->
+                    stream.open(operationContext));
+
+            //then
+            long elapsedMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - connectOpenStart);
+            // Allow for some timing imprecision due to test overhead.
+            int maximumAcceptableTimeoutOvershoot = 300;
+
+            assertInstanceOf(InterruptedByTimeoutException.class, mongoSocketOpenException.getCause());
+            assertFalse(connectTimeoutMs > elapsedMs,
+                    format("Connection timed-out sooner than expected. ConnectTimeoutMS: %d, elapsedMs: %d", connectTimeoutMs, elapsedMs));
+            assertTrue(elapsedMs - connectTimeoutMs < maximumAcceptableTimeoutOvershoot,
```

Comment:
Not that it is important, but given the name "maximumAcceptableTimeoutOvershoot", we should use `<=` instead of `<`. If the "overshoot" is 300, it is still acceptable, right?

```suggestion
            assertTrue(elapsedMs - connectTimeoutMs <= maximumAcceptableTimeoutOvershoot,
```

Reply:
Yes, `<=` aligns with the variable name. Committed, thanks!